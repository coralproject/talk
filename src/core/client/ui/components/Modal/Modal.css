.root {
  position: fixed;
  top: 0px;
  left: 0px;
  display: flex;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  justify-content: center;
  align-items: center;
  z-index: var(--zindex-modal);
}
