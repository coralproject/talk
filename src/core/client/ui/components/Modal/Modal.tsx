import cn from "classnames";
import React, {
  FunctionComponent,
  useCallback,
  useEffect,
  useState,
} from "react";
import ReactDOM from "react-dom";

import { withStyles } from "coral-ui/hocs";
import { PropTypesOf } from "coral-ui/types";

import Backdrop from "../Backdrop";
import NoScroll from "../NoScroll";
import TrapFocus from "../TrapFocus";

import styles from "./Modal.css";

function appendDivNode() {
  const div = document.createElement("div");
  document.body.append(div);
  div.setAttribute("data-portal", "modal");
  return div;
}

/**
 * useDOMNode is a React hook that returns a DOM node
 * to be used as a portal for the modal.
 *
 * @param open whether the modal is open or not.
 */
function useDOMNode(open: boolean) {
  const [modalDOMNode, setModalDOMNode] = useState<HTMLDivElement | null>(null);
  useEffect(() => {
    if (open) {
      const node = appendDivNode();
      setModalDOMNode(node);
      return () => {
        node.parentElement!.removeChild(node);
        setModalDOMNode(null);
      };
    }
    return;
  }, [open]);
  return modalDOMNode;
}

interface Props {
  onClose?: (
    event: React.KeyboardEvent | React.MouseEvent,
    reason: "backdropClick" | "escapeKeyDown"
  ) => void;
  onBackdropClick?: React.EventHandler<React.MouseEvent>;
  onEscapeKeyDown?: React.EventHandler<React.KeyboardEvent>;
  className?: string;
  /**
   * Override or extend the styles applied to the component.
   */
  classes: typeof styles;
  open?: boolean;
  children?: PropTypesOf<typeof TrapFocus>["children"];
}

const Modal: FunctionComponent<Props> = ({
  classes,
  open,
  onClose,
  className,
  onBackdropClick,
  onEscapeKeyDown,
  children,
  ...rest
}) => {
  const rootClassName = cn(classes.root, className);

  const modalDOMNode = useDOMNode(Boolean(open));
  const handleEscapeKeyDown = useCallback(
    (e: React.KeyboardEvent) => {
      if (e.keyCode === 27) {
        if (onEscapeKeyDown) {
          onEscapeKeyDown(e);
        }
        if (onClose) {
          onClose(e, "escapeKeyDown");
        }
      }
    },
    [onEscapeKeyDown, onClose]
  );
  const handleBackdropClick = useCallback(
    (e: React.MouseEvent) => {
      if (onBackdropClick) {
        onBackdropClick(e);
      }
      if (onClose) {
        onClose(e, "backdropClick");
      }
    },
    [onBackdropClick, onClose]
  );
  const handleWrapperClick = useCallback((e: React.MouseEvent) => {
    e.stopPropagation();
  }, []);

  if (open && modalDOMNode) {
    return ReactDOM.createPortal(
      <div role="dialog" className={rootClassName} {...rest}>
        <NoScroll active={open} />
        <Backdrop
          active={open}
          data-testid="backdrop"
          onClick={handleBackdropClick}
        >
          <div
            role="presentation"
            className={styles.scroll}
            onKeyDown={handleEscapeKeyDown}
          >
            <div className={styles.alignContainer1}>
              <div className={styles.alignContainer2}>
                <div
                  className={styles.wrapper}
                  role="presentation"
                  onClick={handleWrapperClick}
                >
                  <TrapFocus>{children}</TrapFocus>
                </div>
              </div>
            </div>
          </div>
        </Backdrop>
      </div>,
      modalDOMNode
    );
  }
  return null;
};

const enhanced = withStyles(styles)(Modal);
export default enhanced;
