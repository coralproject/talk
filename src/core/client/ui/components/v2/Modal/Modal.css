.root {
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  z-index: $zindex-modal;
}

.baseScroll {
  position: relative;

  width: 100%;
  height: 100%;
}

.noScroll {
  overflow-y: hidden;
}

.scroll {
  overflow-y: auto;
}

.alignContainer1 {
  display: table;
  margin: 0 auto;
  height: 100%;
}

.alignContainer2 {
  display: table-cell;
  vertical-align: middle;
}

.wrapper {
  pointer-events: all;
}
