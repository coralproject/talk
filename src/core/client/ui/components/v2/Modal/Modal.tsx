import cn from "classnames";
import React, { FunctionComponent, useCallback } from "react";

import { withStyles } from "coral-ui/hocs";
import { PropTypesOf } from "coral-ui/types";

import Backdrop from "../Backdrop";
import NoScroll from "../NoScroll";
import TrapFocus from "../TrapFocus";

import styles from "./Modal.css";

export interface ModalProps {
  onClose?: (
    event: React.KeyboardEvent | React.MouseEvent,
    reason: "backdropClick" | "escapeKeyDown"
  ) => void;
  onBackdropClick?: React.EventHandler<React.MouseEvent>;
  onEscapeKeyDown?: React.EventHandler<React.KeyboardEvent>;
  className?: string;
  /**
   * Override or extend the styles applied to the component.
   */
  classes: typeof styles;
  open?: boolean;
  children?: PropTypesOf<typeof TrapFocus>["children"];
  disableScroll?: boolean;
}

export const Modal: FunctionComponent<ModalProps> = ({
  classes,
  open,
  onClose,
  className,
  onBackdropClick,
  onEscapeKeyDown,
  children,
  disableScroll = false,
  ...rest
}) => {
  const rootClassName = cn(classes.root, className);

  const handleEscapeKeyDown = useCallback(
    (e: React.KeyboardEvent) => {
      if (e.keyCode === 27) {
        if (onEscapeKeyDown) {
          onEscapeKeyDown(e);
        }
        if (onClose) {
          onClose(e, "escapeKeyDown");
        }
      }
    },
    [onEscapeKeyDown, onClose]
  );
  const handleBackdropClick = useCallback(
    (e: React.MouseEvent) => {
      if (e.target !== e.currentTarget) {
        return;
      }
      if (onBackdropClick) {
        onBackdropClick(e);
      }
      if (onClose) {
        onClose(e, "backdropClick");
      }
    },
    [onBackdropClick, onClose]
  );

  if (open) {
    const content = (
      <div role="dialog" className={rootClassName} {...rest}>
        <NoScroll active={open} />
        <Backdrop active={open} />
        <div
          role="presentation"
          className={cn(
            styles.baseScroll,
            disableScroll ? styles.noScroll : styles.scroll
          )}
          onKeyDown={handleEscapeKeyDown}
          onClick={handleBackdropClick}
          data-testid="scroll"
        >
          <div className={styles.alignContainer1}>
            <div className={styles.alignContainer2}>
              <div className={styles.wrapper}>
                <TrapFocus>{children}</TrapFocus>
              </div>
            </div>
          </div>
        </div>
      </div>
    );

    return content;
  }
  return null;
};

const enhanced = withStyles(styles)(Modal);
export default enhanced;
