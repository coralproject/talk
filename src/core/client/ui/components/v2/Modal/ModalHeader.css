$modal-header-background: $colors-white-500;

.root {
  background-color: $modal-header-background;
  padding: var(--spacing-4);
}
