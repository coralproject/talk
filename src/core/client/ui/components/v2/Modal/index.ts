export { default, ModalProps } from "./Modal";
export { default as ModalHeader } from "./ModalHeader";
