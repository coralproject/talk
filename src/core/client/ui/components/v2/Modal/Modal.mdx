---
name: Modal
menu: UI Kit
---

import { Playground, PropsTable } from "docz";
import { useState, useCallback } from "react";

import Button from "../Button";
import Card, { CardCloseButton } from "../Card";
import Flex from "../Flex";
import HorizontalGutter from "../HorizontalGutter";
import Modal from "./Modal";

import styles from "./Modal.mdx.css";

# Modal

## Basic usage

<Playground>
  {() => {
    const [open, setOpen] = useState(false);
    const show = useCallback(() => setOpen(true));
    const hide = useCallback(() => setOpen(false));
    return (
      <>
        <Button variant="filled" color="primary" onClick={show}>
          Show Modal
        </Button>
        <Modal open={open} onClose={hide} aria-labelledby="modal-title">
          {({ firstFocusableRef, lastFocusableRef }) => (
            <Card className={styles.card}>
              <CardCloseButton onClick={hide} ref={firstFocusableRef} />
              <HorizontalGutter size="double">
                <HorizontalGutter>
                  <h2 id="modal-title">
                    Are you sure you want to ban <strong>ButFirstCoffee</strong>?
                  </h2>
                  <Typography>
                    Once banned, this user will no longer be able to comment,
                    use reactions, or report comments.
                  </Typography>
                </HorizontalGutter>
                <Flex justifyContent="flex-end" itemGutter="half">
                  <Button variant="outlined" onClick={hide}>
                    Cancel
                  </Button>
                  <Button
                    variant="filled"
                    color="primary"
                    onClick={hide}
                    ref={lastFocusableRef}
                  >
                    Ban User
                  </Button>
                </Flex>
              </HorizontalGutter>
            </Card>
          )}
        </Modal>
      </>
    );
  }}
</Playground>
