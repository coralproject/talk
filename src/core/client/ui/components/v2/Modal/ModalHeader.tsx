import React, { FunctionComponent, RefObject } from "react";

import { CardCloseButton, Flex } from "coral-ui/components/v2";

import styles from "./ModalHeader.css";

interface Props {
  onClose: () => void;
  focusableRef?: RefObject<any>;
  children?: React.ReactNode;
}

const ModalHeader: FunctionComponent<Props> = ({
  onClose,
  focusableRef,
  children,
}) => {
  return (
    <div className={styles.root}>
      <Flex justifyContent="space-between" alignItems="flex-start">
        {children}
        <CardCloseButton onClick={onClose} ref={focusableRef} />
      </Flex>
    </div>
  );
};

export default ModalHeader;
