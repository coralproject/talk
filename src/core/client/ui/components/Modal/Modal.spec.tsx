import { noop } from "lodash";
import React from "react";
import TestRenderer, { ReactTestRenderer } from "react-test-renderer";
import sinon from "sinon";

import Modal from "./Modal";

const createNodeMock = () => ({
  focus: noop,
});

it("renders correctly", () => {
  expect(document.body.lastChild).toBeNull();
  let testRenderer: ReactTestRenderer;
  TestRenderer.act(() => {
    testRenderer = TestRenderer.create(
      <Modal>
        <div>Test</div>
      </Modal>,
      { createNodeMock }
    );
  });
  expect(testRenderer!.toJSON()).toBeNull();
  expect(document.body.lastChild).toBeNull();
  TestRenderer.act(() => {
    testRenderer.update(
      <Modal open>
        <div>Test</div>
      </Modal>
    );
  });
  expect(testRenderer!.toJSON()).toMatchSnapshot();
  expect(document.body.lastElementChild!.getAttribute("data-portal")).toBe(
    "modal"
  );
  TestRenderer.act(() => {
    testRenderer.unmount();
  });
  expect(document.body.lastChild).toBeNull();
});

it("relays backdrop click events", () => {
  const event = {};
  const onBackdropClick = sinon.stub();
  const onClose = sinon.stub();
  expect(document.body.lastChild).toBeNull();
  let testRenderer: ReactTestRenderer;
  TestRenderer.act(() => {
    testRenderer = TestRenderer.create(
      <Modal open onBackdropClick={onBackdropClick} onClose={onClose}>
        <div>Test</div>
      </Modal>,
      { createNodeMock }
    );
  });
  testRenderer!.root
    .findByProps({ "data-testid": "backdrop" })
    .props.onClick(event);
  TestRenderer.act(() => {
    testRenderer.unmount();
  });
  expect(onBackdropClick.called).toBe(true);
  expect(onClose.calledWith(event, "backdropClick")).toBe(true);
});

it("relays esc events", () => {
  const escEvent = { keyCode: 27 };
  const otherEvent = { keyCode: 100 };
  const onEscapeKeyDown = sinon.stub();
  const onClose = sinon.stub();
  expect(document.body.lastChild).toBeNull();
  let testRenderer: ReactTestRenderer;
  TestRenderer.act(() => {
    testRenderer = TestRenderer.create(
      <Modal open onEscapeKeyDown={onEscapeKeyDown} onClose={onClose}>
        <div>Test</div>
      </Modal>,
      { createNodeMock }
    );
  });
  const el = testRenderer!.root.find(i => i.props.onKeyDown);
  el.props.onKeyDown(escEvent);
  el.props.onKeyDown(otherEvent);
  TestRenderer.act(() => {
    testRenderer.unmount();
  });
  expect(onEscapeKeyDown.called).toBe(true);
  expect(onClose.calledWith(escEvent, "escapeKeyDown")).toBe(true);
  expect(onClose.calledWith(otherEvent, "escapeKeyDown")).toBe(false);
});
