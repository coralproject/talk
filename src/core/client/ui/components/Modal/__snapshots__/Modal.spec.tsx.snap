// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Modal-root"
  onKeyDown={[Function]}
  role="modal"
>
  <div
    className="Backdrop-root Backdrop-active"
    data-testid="backdrop"
    onClick={[Function]}
  />
  <div
    onFocus={[Function]}
    tabIndex={0}
  />
  <div
    tabIndex={-1}
  />
  <div>
    Test
  </div>
  <div
    onFocus={[Function]}
    tabIndex={0}
  />
</div>
`;
